File: openmodels/registry.py
from __future__ import annotations

from .adapter import AdapterError, BackendAdapter
from .drizzle import DRIZZLE_PG_ADAPTER


ADAPTERS: dict[str, BackendAdapter] = {
    DRIZZLE_PG_ADAPTER.key: DRIZZLE_PG_ADAPTER,
}


def get_adapter(target: str) -> BackendAdapter:
    try:
        return ADAPTERS[target]
    except KeyError as exc:
        supported = ", ".join(sorted(ADAPTERS))
        raise AdapterError(
            f"Unknown adapter target '{target}'. Supported targets: {supported}."
        ) from exc


def list_adapters() -> list[BackendAdapter]:
    return [ADAPTERS[key] for key in sorted(ADAPTERS)]

File: rust/openmodels-rs/src/registry.rs
use crate::adapter::BackendAdapter;
use crate::drizzle::{DRIZZLE_PG_ADAPTER, DRIZZLE_PG_TARGET};
use crate::error::{message, Result};
use crate::seaorm::{SEAORM_RUST_ADAPTER, SEAORM_RUST_TARGET};

pub fn get_adapter(target: &str) -> Result<&'static dyn BackendAdapter> {
    match target {
        DRIZZLE_PG_TARGET => Ok(&DRIZZLE_PG_ADAPTER),
        SEAORM_RUST_TARGET => Ok(&SEAORM_RUST_ADAPTER),
        _ => Err(message(format!("Unknown adapter target '{}'.", target))),
    }
}

pub fn list_adapters() -> Vec<&'static dyn BackendAdapter> {
    vec![&DRIZZLE_PG_ADAPTER, &SEAORM_RUST_ADAPTER]
}
